add pic18 ltt uart driver

AndrewWang001 opened the pull request:
## Added 
- pic18f26k83/uart.c
- include/uart.h 
(from ltt 2024 code)

## Testing
1. **Wiring**
- Flash: Computer -> Snap -> LTT Board (MCLR etc)
- Power: 12V & Ground -> Power Supply
- TTL to USB Serial Converter (datasheet: https://www.ftdichip.com/Support/Documents/DataSheets/Cables/DS_[TTL-232R](https://www.ftdichip.com/Support/Documents/DataSheets/Cables/DS_TTL-232R_CABLES.pdf)_CABLES.pdf)
  - PIC18 GND -> USB-Serial GND (black)
  - PIC18 CTS -> USB-Serial CTS# (brown)
  - PIC18 RX -> USB-Serial RXD (yellow)
  - PIC18 TX -> USB-Serial TXD (orange)
  - PIC18 RTS -> USB-Serial RTS# (green)
2. **Software**
- switched to ltt repo to test uart
- used ltt's main as is, except
  - commented out all CAN/radio logic to remove console spam:
       ```c
       // radio current checks
       // can_msg_t msg;
       // …
       ```
  - Inserted in loop:
       ```c
       uint8_t test[] = "Hello World ";
       uart_transmit_buffer(test, sizeof(test) - 1);
       ```
- Build and Flash
3. **Run**
- Open PuTTY on COM port for FTDI cable (idk mac)
- 115200 baud rate
- should see text Hello World Hello World Hello World … printing every 500ms

![Screenshot 2025-05-23 015030](https://github.com/user-attachments/assets/a3ffca3e-3789-41d2-8e36-4c639af28764)
![Screenshot 2025-05-23 015127](https://github.com/user-attachments/assets/295d8541-4ea6-414b-96d8-72df93b49982)
![Screenshot 2025-05-23 015147](https://github.com/user-attachments/assets/65843fa3-53ab-43e7-9aaa-1ed0cf65a3de)

JasonBrave commented on include/uart.h:
```diff
@@ -0,0 +1,42 @@
+#ifndef UART_H_
```
```suggestion
#ifndef ROCKETLIB_UART_H
```

JasonBrave commented on include/uart.h:
```diff
@@ -0,0 +1,42 @@
+#ifndef UART_H_
+#define UART_H_
+
+#include "common.h"
```
put common.h after standard C heaters, separate by an empty line

JasonBrave commented on pic18f26k83/uart.c:
```diff
@@ -0,0 +1,139 @@
+// memory pools to use for those srbs. 100 is a completely arbitrary number
+uint8_t rx_buffer_pool[100], tx_buffer_pool[100];
+
+w_status_t uart_init(uint32_t baud_rate, uint32_t fosc, bool enable_flow_control)
```
syntax error bruh( missing { )

AndrewWang001 commented:
### UART driver tested on LTT hardware and verified working.

**Steps taken:**
- Switched to LTT codebase (`uart.h`, `uart.c`, and `main.c`)
- Replaced `uart.h` with the new version and inlined `common.h` to define `w_status_t` (since LTT didn’t include it)
- Replaced `uart.c` with the new version and **re-added PPS pin configuration** (was previously removed for external usage)
- Modified `main.c` to use the updated initialization call:  
  `uart_init(115200, 48000000UL, true);`
- Flashed firmware and confirmed proper UART output in PuTTY at 115200 baud with flow control enabled

Correct output of `"Hello World"` repeating every 500ms — also noticed the weird garbage characters from early UART startup are now gone.

![image](https://github.com/user-attachments/assets/dc9d2703-a972-4b63-a799-f76a341cd322)

AndrewWang001 commented:
Tested with different baud rates

JasonBrave commented on pic18f26k83/uart.c:
```diff
@@ -0,0 +1,122 @@
+#include <pic18f26k83.h>
```
Do not include this file #include <pic18f26k83.h>, include xc8.h instead

JasonBrave commented on pic18f26k83/uart.c:
```diff
@@ -0,0 +1,122 @@
+#include <stdbool.h>
+#include <stdint.h>
+
+#include "canlib.h"
```
We don't want canlib dependecy

JasonBrave commented on pic18f26k83/uart.c:
```diff
@@ -0,0 +1,122 @@
+#include "common.h"
+#include "uart.h"
+
+// safe ring buffers for sending and receiving
```
Get rid of buffer

JasonBrave commented on pic18f26k83/uart.c:
```diff
@@ -0,0 +1,122 @@
+    // we are go for UART
+    U1CON1bits.ON = 1;
+
+    // initialize the rx and tx buffers
```
Do not use buffer

JasonBrave commented on pic18f26k83/uart.c:
```diff
@@ -0,0 +1,122 @@
+    }
+}
+
+bool uart_byte_available(void) {
```
delete this

JasonBrave commented on pic18f26k83/uart.c:
```diff
@@ -0,0 +1,122 @@
+        tx++;
+    }
+    // If the module isn't enabled, give it a byte to send and enable it
+    if (PIE3bits.U1TXIE == 0) {
```
Make this into a uart transmit byte function

JasonBrave commented on pic18f26k83/uart.c:
```diff
@@ -0,0 +1,122 @@
+    return !srb_is_empty(&rx_buffer);
+}
+
+uint8_t uart_read_byte(void) {
```
Get rid of this

JasonBrave commented on pic18f26k83/uart.c:
```diff
@@ -0,0 +1,122 @@
+    return rcv;
+}
+
+void uart_interrupt_handler(void) {
```
For interrupt handler, for receive, implement a callback option

JasonBrave commented on pic18f26k83/uart.c:
```diff
@@ -0,0 +1,122 @@
+    return W_SUCCESS;
+}
+
+void uart_transmit_buffer(uint8_t *tx, uint8_t len) {
```
Implement transmit buffer without canlib dependency
